src/compile/BuildEnv.zig

```diff
@@ -0,0 +1,2702 @@
+    defer gpa.free(app_path);
+
+    // Build should complete without deadlocks or race conditions
+    try ws.buildApp(app_path);
+
+    // Verify no duplicate processing occurred by checking phase progression
+    const app_build = ws.packages.get("app").?;
+    const main_sched = app_build.schedulers.get("app").?;
+
+    main_sched.lock.lock();
+    defer main_sched.lock.unlock();
+
+    // All modules should reach Done phase exactly once
+    var it = main_sched.modules.iterator();
+    while (it.next()) |entry| {
+        const state = entry.value_ptr.*;
+        try std.testing.expectEqual(PackageEnv.Phase.Done, state.phase);
+        const working_val = if (@import("builtin").target.cpu.arch != .wasm32) state.working.load(.seq_cst) else state.working;
+        try std.testing.expectEqual(@as(u8, 0), working_val);
+    }
+
+    // Drain reports (may have cycle errors due to circular imports)
+    const drained = try ws.drainReports();
+    defer {
+        var idx: usize = 0;
+        while (idx < drained.len) : (idx += 1) {
+            for (drained[idx].reports) |*rep| rep.deinit();
+            gpa.free(drained[idx].reports);
+            gpa.free(drained[idx].abs_path);
+        }
+        gpa.free(drained);
+    }
+}
+
+test "BuildEnv: streaming with module chain" {
+    const gpa = std.testing.allocator;
+    var tmp = try std.testing.tmpDir(.{});
+    defer tmp.cleanup();
+
+    const root_dir = tmp.dir.path;
+    try tmp.dir.makePath("app");
+
+    // Create a chain of modules A -> B -> C -> D
+    try writeFile(tmp.dir, "app/D.roc",
+        \\module [valueD]
+        \\
+        \\valueD = 4
+    );
+
+    try writeFile(tmp.dir, "app/C.roc",
+        \\module [valueC]
+        \\
+        \\import D
+        \\
+        \\valueC = D.valueD + 3
+    );
+
+    try writeFile(tmp.dir, "app/B.roc",
+        \\module [valueB]
+        \\
+        \\import C
+        \\
+        \\valueB = C.valueC + 2
+    );
+
+    try writeFile(tmp.dir, "app/A.roc",
+        \\module [valueA]
+        \\
+        \\import B
+        \\
+        \\valueA = B.valueB + 1
+    );
+
+    try writeFile(tmp.dir, "app/Main.roc",
+        \\app [main!] {}
+        \\
+        \\import A
+        \\
+        \\main! = |_| A.valueA
+    );
+
+    // Use single-threaded mode for deterministic testing
+    var ws = BuildEnv.init(gpa, .single_threaded, 1);
+    defer ws.deinit();
+
+    const app_path = try std.fs.path.join(gpa, &.{ root_dir, "app", "Main.roc" });
+    defer gpa.free(app_path);
+
+    // Build the app normally
+    try ws.buildApp(app_path);
+
+    // Should have successfully built all modules
+    // Verify by checking that we can drain reports
+    const drained = try ws.drainReports();
+    defer cleanupReports(gpa, drained);
+
+    // Expect no errors from this simple module chain
+    try std.testing.expectEqual(@as(usize, 0), drained.len);
+}
+
+test "BuildEnv: enforce rules (only apps -> platforms, no package -> app)" {
```

Comment:
None of these tests are running with `zig build test`... I'll move them into a separate file